### ISSUE TITLE
CoreDNS, Kubernetes, and systemd - CrashLoopBackOff

### ISSUE THREAD

**Issue by buckaroogeek** [miekg/coredns#6586]
I have been testing instantiation of small Kubernetes clusters using kubeadm on libvirt virtual machines. When the cluster is initiated by kubeadm, coredns pods  inevitably have a status of CrashLoopBackOff triggered by loop detection:

```
[bgsmith@testf39 vagrant]$ kubectl get pods -A
NAMESPACE      NAME                              READY   STATUS             RESTARTS        AGE
kube-flannel   kube-flannel-ds-85qzj             1/1     Running            0               10m
kube-system    coredns-5dd5756b68-ds6cj          0/1     CrashLoopBackOff   6 (3m52s ago)   10m
kube-system    coredns-5dd5756b68-gnhp9          0/1     CrashLoopBackOff   6 (4m3s ago)    10m
kube-system    etcd-testf39                      1/1     Running            0               10m
kube-system    kube-apiserver-testf39            1/1     Running            0               10m
kube-system    kube-controller-manager-testf39   1/1     Running            0               10m
kube-system    kube-proxy-7bjxd                  1/1     Running            0               10m
kube-system    kube-scheduler-testf39            1/1     Running            0               10m
```

I have to then edit the coredns configmap and change
```
        forward . /etc/resolv.conf {
           max_concurrent 1000
        }

```
to
```
        forward . 192.168.121.1 {
           max_concurrent 1000
        }
```

where 192.168.121.x is the subnet libvirt uses as the default for nodes and 192.168.121.1 is the nameserver/gateway.
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:6f:f0:50 brd ff:ff:ff:ff:ff:ff
    altname enp0s5
    altname ens5
    inet 192.168.121.192/24 brd 192.168.121.255 scope global dynamic noprefixroute eth0
       valid_lft 3345sec preferred_lft 3345sec
    inet6 fe80::3b74:b8c9:252e:5f0/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```
When I look at the kubelet configuration that kubeadm writes to ```/var/lib/kubelet/config.yaml``` I can see that the value for resolveConf is correct for a node using systemd like these virtual machines are (Fedora 39 cloud image):
```
resolvConf: /run/systemd/resolve/resolv.conf
```

Why does CoreDNS use ```/etc/resolv.conf``` instead of the dns resolver that the kubelet is configured for? Is there a way I can configure CoreDNS to use the same value that the kubelet does? 

The loop troubleshooting section (https://coredns.io/plugins/loop/#troubleshooting) seems to identify the problem correctly but the recommended solution (add a kubelet resolvConf configuration for  /run/systemd/resolve/resolv.conf) is already correctly set so I am puzzled.

I suspect this is the 10,000th time this question has been asked but I cannot seem to find the correct solution so would appreciate any pointers. 

many thanks

**Comment by chrisohaver** [comment]
> Why does CoreDNS use /etc/resolv.conf instead of the dns resolver that the kubelet is configured for? Is there a way I can configure CoreDNS to use the same value that the kubelet does?

It doesn’t actually use the same file. The /etc/resolv.conf Coredns uses is relative to the Pod coredns runs in. This is not necessarily the same as the host’s /etc/resolv.conf. kubelet sets up pods to use the resolv.conf that it’s configured with (the resolvConf setting).

**Comment by buckaroogeek** [comment]
Thank you for the information. Much appreciated.  I have  a lot to learn!

**Comment by buckaroogeek** [comment]
And closing ...

**Comment by Erokos** [comment]
Hi, I know I'm commenting on the same issue but @buckaroogeek have you managed to solve this? I've created an EKS cluster using `eksctl` and am stuck with this. The nodes use Amazon linux 2 which is the same family as Fedora and systemd is used but there is no file at `/run/systemd/resolve/resolv.conf` and I don't have this configured in `/etc/kubernetes/kubelet/kubelet-config.json`.
@chrisohaver do you maybe have some insight into this issue? Comparing to an older cluster running on Amazon Linux 2 with docker as the container runtime and cgroups as the driver, I don't have this issue.

**Comment by buckaroogeek** [comment]
I can add an update. My specific configuration is Fedora with Kubernetes and CRI-O rpms from Fedora repositories using kubeadm to initialize clusters. Older Kubernetes rpms in Fedora depend on command line parameters to configure kubelet (via kubeadm initialization). It was this configuration where experiences the CrashLoopBackOff result. In this configuration I wrote up a note on one solution at: https://docs.fedoraproject.org/en-US/quick-docs/using-kubernetes-kubeadm/#_troubleshooting_crashloopbackoff.

However, more current Kuberenetes rpms in Fedora now default to using the kubelet configuration file. With the default kubelet configuration file, I found I no longer experienced CrashLoopBackOff results in cluster set up.

I speculate that the default configuration file for the kubelet has some setting that avoids this result, but I have not tried to explicitly debug as the older Kubernetes rpms with the command-line configuration are approaching end of life in Fedora. All versioned Kubernetes rpms (e.g. kubernetes1.29 and kubernetes1.31, etc) now in Fedora 41 and newer (or from COPR for F40 or F39) use the configuration file.
